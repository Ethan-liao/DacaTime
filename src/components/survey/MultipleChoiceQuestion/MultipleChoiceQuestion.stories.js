import React from 'react';
import { storiesOf, action, linkTo } from '@kadira/storybook';
import { component as MultipleChoiceQuestion } from './index';
import { loremIpsum } from '../../../stories/util';

storiesOf('Multiple choice question (radio)', module)
  .add('two choices, no explanation', () =>
    <MultipleChoiceQuestion
      questionId={1}
      name="mc-question"
      prompt="What's your favorite side of a coin?"
      fields={[
        { label: 'Heads', name: 'heads' },
        { label: 'Tails', name: 'tails' },
      ]}
      dispatch={action('updated-field')}
    />,
  )
  .add('two choices, with explanation', () =>
    <MultipleChoiceQuestion
      questionId={1}
      name="mc-question"
      prompt="What's your favorite side of a coin?"
      fields={[
        { label: 'Heads', name: 'heads' },
        { label: 'Tails', name: 'tails' },
      ]}
      explanation="Flipping coins is very complicated, so let me explain it to you..."
      dispatch={action('updated-field')}
    />,
  )
  .add('five choices, with explanation', () =>
    <MultipleChoiceQuestion
      questionId={1}
      name="mc-question"
      prompt="What's your favorite protein?"
      fields={[
        { label: 'Chicken', name: 'chicken' },
        { label: 'Beef', name: 'beef' },
        { label: 'Turkey', name: 'turkey' },
        { label: 'Soy', name: 'soy' },
        { label: 'Pork', name: 'Pork' },
      ]}
      explanation={loremIpsum}
      dispatch={action('updated-field')}
    />,
  );
