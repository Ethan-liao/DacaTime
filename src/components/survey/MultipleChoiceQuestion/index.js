import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import QuestionContainer from '../QuestionContainer';
import { updateField } from '../../../actions';
import RadioButton from '../../basic/RadioButton';

const changeHandler = dispatch => targetKeyPath => event => {
  const value = event.target.value;
  dispatch(updateField(targetKeyPath, value));
};

const MultipleChoiceQuestion = props => {
  const updateKeyPath = [props.name];
  const value = props.value || '';

  const fields = props.fields.map(({ label, name }) =>
    <RadioButton
      key={name}
      name={name}
      label={label}
      questionId={props.questionId}
      value={value}
    />,
  );

  const className =
    fields.length > 2
      ? 'dt-question-container-vertical'
      : 'dt-question-container-horizontal';

  return (
    <QuestionContainer {...props}>
      <div
        className={className}
        value={value}
        onChange={changeHandler(props.dispatch)(updateKeyPath)}
      >
        {fields}
      </div>
    </QuestionContainer>
  );
};
MultipleChoiceQuestion.propTypes = {
  questionId: PropTypes.number.isRequired,
  prompt: PropTypes.string.isRequired,
  explanation: PropTypes.string,
  fields: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
    }),
  ).isRequired,
  dispatch: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  dispatch,
});

const mapStateToProps = (state, ownProps) => ({
  value: state.form[ownProps.name],
  ...ownProps,
});

export const component = MultipleChoiceQuestion;
export default connect(mapStateToProps, mapDispatchToProps)(component);
