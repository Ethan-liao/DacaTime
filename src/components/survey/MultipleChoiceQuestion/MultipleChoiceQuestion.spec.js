import React from 'react';
import MultipleChoiceQuestion from './index';
import renderer from 'react-test-renderer';
import { radioQuestion } from '../../../util/testData';
import reducer from '../../../reducers';
import { setup } from '../../../util/testUtils';

const { component, store } = setup(
  MultipleChoiceQuestion,
  reducer,
  radioQuestion,
);

it('renders without crashing', () => {
  renderer.create(component);
});

it('renders correctly', () => {
  const renderedComponent = renderer.create(component);
  let tree = renderedComponent.toJSON();

  expect(tree).toMatchSnapshot();
});
