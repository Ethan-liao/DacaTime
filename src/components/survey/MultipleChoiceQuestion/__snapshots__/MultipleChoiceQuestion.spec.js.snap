exports[`test renders correctly 1`] = `
<div
  className="dt-panel dt-question-container">
  <div
    className="dt-question-header">
    Were you under the age of 31 as of June 15, 2012?
  </div>
  <div
    className="dt-question-container-horizontal"
    onChange={[Function]}
    value="">
    <div
      className="dt-radio-container">
      <input
        checked={false}
        className="dt-radio-input"
        hidden={true}
        id="radio-1-yes"
        name="radio-1-yes"
        onChange={[Function]}
        type="radio"
        value="yes" />
      <label
        className="dt-radio-label"
        htmlFor="radio-1-yes">
        Yes
      </label>
      <div
        className="dt-radio-button" />
    </div>
    <div
      className="dt-radio-container">
      <input
        checked={false}
        className="dt-radio-input"
        hidden={true}
        id="radio-1-no"
        name="radio-1-no"
        onChange={[Function]}
        type="radio"
        value="no" />
      <label
        className="dt-radio-label"
        htmlFor="radio-1-no">
        No
      </label>
      <div
        className="dt-radio-button" />
    </div>
  </div>
</div>
`;
